src/Microsoft.OData.Core/JsonLight/ODataJsonLightOutputContext.cs

```diff
@@ -953,32 +945,73 @@ protected override void Dispose(bool disposing)
             IJsonWriterAsync asynchronousJsonWriter;
             if (container == null)
             {
-                if (this.concreteJsonWriter != null)
-                {
-                    return this.concreteJsonWriter;
-                }
+                // we don't create a default JsonWriter here because that's already handled by the CreateJsonWriter() method.
+                return null;
+            }
+
+            IJsonWriterFactoryAsync asynchronousJsonWriterFactory = container.GetRequiredService<IJsonWriterFactoryAsync>();
+            asynchronousJsonWriter = asynchronousJsonWriterFactory.CreateAsynchronousJsonWriter(textWriter, isIeee754Compatible);
+            Debug.Assert(asynchronousJsonWriter != null, "asynchronousJsonWriter != null");
 
-                asynchronousJsonWriter = new JsonWriter(textWriter, isIeee754Compatible);
+            if (asynchronousJsonWriter is Json.JsonWriter defaultJsonWriter)
+            {
+                defaultJsonWriter.ArrayPool = writerSettings.ArrayPool;
             }
-            else
+
+            return asynchronousJsonWriter;
+        }
+
+        private IJsonWriterAsync CreateAsynchronousJsonWriter(
+            IStreamBasedJsonWriterFactory factory,
+            Stream outputStream,
+            bool isIeee754Compatible,
+            Encoding encoding)
+        {
+            IJsonWriterAsync jsonWriter = factory.CreateAsynchronousJsonWriter(outputStream, isIeee754Compatible, encoding);
+
+            if (jsonWriter == null)
             {
-                IJsonWriterFactoryAsync asynchronousJsonWriterFactory = container.GetRequiredService<IJsonWriterFactoryAsync>();
-                if (asynchronousJsonWriterFactory is DefaultJsonWriterFactory && this.concreteJsonWriter != null)
-                {
-                    return this.concreteJsonWriter;
-                }
+                throw new ODataException(Strings.ODataMessageWriter_StreamBasedJsonWriterFactory_ReturnedNull(encoding.WebName, isIeee754Compatible));
+            }
 
-                asynchronousJsonWriter = asynchronousJsonWriterFactory.CreateAsynchronousJsonWriter(textWriter, isIeee754Compatible);
-                Debug.Assert(asynchronousJsonWriter != null, "asynchronousJsonWriter != null");
+            return jsonWriter;
+        }
+
+        private void VerifyJsonWriter()

```

Comment:
Is `VerifyJsonWriter` a good fit for this method given that it does a lot more than verify?

Reply:
No it's not. Not sure how to name it to reflect what it does, that could be a sign that it does too much 🥲

Reply:
adding a comment of what it does can also help

Reply:
Renamed to `EnsureJsonWritersAreTheSame`